io-engine/src/lvs/lvs_lvol.rs

```diff
@@ -1192,33 +1193,52 @@ impl SnapshotOps for Lvol {
             snap_uuid,
         ))
     }
-    /// List Replica Snapshot.
+    /// List Snapshot
     fn list_snapshot(&self) -> Vec<VolumeSnapshotDescriptor> {
         let mut snapshot_list: Vec<VolumeSnapshotDescriptor> = Vec::new();
-        let bdev =
-            UntypedBdev::bdev_first().expect("Failed to enumerate devices");
-
-        let lvol_devices = bdev
-            .into_iter()
-            .filter(|b| b.driver() == "lvol")
-            .map(|b| Lvol::try_from(b).unwrap())
-            .collect::<Vec<Lvol>>();
-        // return empty snapshot parameter list, if no snapshot found.
-        if lvol_devices.len() <= 1 {
-            return snapshot_list;
-        }
-        for snapshot_lvol in lvol_devices {
-            // skip lvol if it is not snapshot.
-            if !snapshot_lvol.is_snapshot() {
-                continue;
+        if let Some(bdev) = UntypedBdev::bdev_first() {
+            let lvol_devices = bdev
+                .into_iter()
+                .filter(|b| b.driver() == "lvol")
+                .map(|b| Lvol::try_from(b).unwrap())
+                .collect::<Vec<Lvol>>();
+            // return empty snapshot parameter list, if no snapshot found.
+            if lvol_devices.len() <= 1 {
+                return snapshot_list;
             }
-            match snapshot_lvol.snapshot_descriptor(Some(self)) {
-                Some(snapshot_descriptor) => {
-                    snapshot_list.push(snapshot_descriptor)
+            for snapshot_lvol in lvol_devices {
+                // skip lvol if it is not snapshot.
+                if !snapshot_lvol.is_snapshot() {
+                    continue;
+                }
+                match snapshot_lvol.snapshot_descriptor(Some(self)) {
+                    Some(snapshot_descriptor) => {
+                        snapshot_list.push(snapshot_descriptor)
+                    }
+                    None => continue,
                 }
-                None => continue,
             }
-        }
+        };
+        snapshot_list
+    }
+    /// List Single snapshot details based on UUID.
+    fn list_single_snapshot(&self) -> Vec<VolumeSnapshotDescriptor> {
```

Comment:
hmm, seems like these list snapshots won't work if the base replica lvol is deleted?

Reply:
Yes. If replica is deleted, then this API will not return the snapshot result as the snapshot property stored in blobstore will be deleted as part of replica deletion. The other ListSnapshot API (list_all_snapshot) will return the snapshot with valid flag as `false`

Reply:
Why is valid false after deleting the replica?

Reply:
> Why is valid false after deleting the replica?

During listing snapshot, snapshot parent-id (xattr) is validated. As the xattr (parent_uuid -> source_uuid) is not valid, the snapshot is marked as invalid.
But I think I can remove this check, as snapshot can be used after replica deletion. So making `valid_flag` false is not correct.

Reply:
xattr information in the blobstor is valid. But as the replica is deleted already, this information will misslead. So We are returning the result as `null` and `0` for `source_uuid` and `source_size` as part of ListSnapshot response.

Reply:
Why would it mislead? 
Anyways, this means we have to make these optional across the stack then, CC @Abhinandan-Purkait

Reply:
The assumption here is ListSnapshot should not response any parameter which is invalid in the system. As replica is already deleted, DP should not expose that UUID to CP.